File: apps/frontend/src/api/knowledgeGraph.ts
// 知识图谱 API 客户端

import client from './client';

export interface GraphNode {
  id: number;
  name: string;
  node_type: 'concept' | 'topic' | 'entity';
  description?: string;
  source_note_ids?: string[];
}

export interface GraphEdge {
  id: number;
  source_node_id: number;
  target_node_id: number;
  edge_type: 'related' | 'hierarchical' | 'sequential';
  description?: string;
  strength: number;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface KnowledgeGraphStatus {
  has_llm_config: boolean;
  graph_exists: boolean;
  node_count: number;
  edge_count: number;
  is_building: boolean;
  building_progress: number;
  is_full_build?: boolean; // 当前进行中的任务是否为全量构建
  last_built_at?: string;
}

export interface BuildGraphProgress {
  status: 'idle' | 'pending' | 'running' | 'completed' | 'failed';
  progress: number;
  message: string;
  stats?: Record<string, any>;
}

export const knowledgeGraphApi = {
  // 获取图谱状态
  getStatus: async (): Promise<KnowledgeGraphStatus> => {
    const response = await client.get('/knowledge-graph/status');
    return response.data;
  },

  // 获取图谱数据
  getData: async (): Promise<GraphData> => {
    const response = await client.get('/knowledge-graph/data');
    return response.data;
  },

  // 构建图谱（创建任务）
  buildGraph: async (rebuild: boolean = false): Promise<any> => {
    console.log('[knowledgeGraphApi] buildGraph 调用, rebuild:', rebuild);
    const response = await client.post('/tasks', {
      task_type: 'knowledge_graph_build',
      title: rebuild ? '重新构建知识图谱' : '构建知识图谱',
      description: rebuild ? '清空现有数据并重新构建知识图谱' : '从笔记中提取概念构建知识图谱',
      metadata_json: JSON.stringify({ rebuild }),
    });
    console.log('[knowledgeGraphApi] buildGraph 响应:', response.data);
    return response.data;
  },

  // 获取构建进度（通过任务详情获取）
  getBuildProgress: async (taskId: number): Promise<any> => {
    const response = await client.get(`/tasks/${taskId}`);
    return response.data;
  },

  // 清空图谱
  clearGraph: async (): Promise<void> => {
    await client.delete('/knowledge-graph/data');
  },

  // 获取所有节点
  getNodes: async (): Promise<GraphNode[]> => {
    const response = await client.get('/knowledge-graph/nodes');
    return response.data;
  },

  // 获取所有边
  getEdges: async (): Promise<GraphEdge[]> => {
    const response = await client.get('/knowledge-graph/edges');
    return response.data;
  },

  // 删除节点
  deleteNode: async (nodeId: number): Promise<void> => {
    await client.delete(`/knowledge-graph/nodes/${nodeId}`);
  },

  // 删除边
  deleteEdge: async (edgeId: number): Promise<void> => {
    await client.delete(`/knowledge-graph/edges/${edgeId}`);
  },
};

File: apps/backend/app/schemas/knowledge_graph.py
# 知识图谱 Schema

from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from datetime import datetime
from enum import Enum


class NodeTypeEnum(str, Enum):
    CONCEPT = "concept"
    TOPIC = "topic"
    ENTITY = "entity"


class EdgeTypeEnum(str, Enum):
    RELATED = "related"
    HIERARCHICAL = "hierarchical"
    SEQUENTIAL = "sequential"


class GraphNodeBase(BaseModel):
    name: str
    node_type: NodeTypeEnum = NodeTypeEnum.CONCEPT
    description: Optional[str] = None
    source_note_ids: Optional[str] = None


class GraphNodeCreate(GraphNodeBase):
    user_id: int


class GraphNodeResponse(GraphNodeBase):
    id: int
    user_id: int
    is_verified: bool = False
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class GraphEdgeBase(BaseModel):
    source_node_id: int
    target_node_id: int
    edge_type: EdgeTypeEnum = EdgeTypeEnum.RELATED
    description: Optional[str] = None
    strength: float = 0.5


class GraphEdgeCreate(GraphEdgeBase):
    user_id: int


class GraphEdgeResponse(GraphEdgeBase):
    id: int
    user_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class GraphData(BaseModel):
    """图谱数据（用于前端可视化）"""
    nodes: List[Dict[str, Any]]
    edges: List[Dict[str, Any]]


class KnowledgeGraphStatus(BaseModel):
    """知识图谱静态状态（不包含构建任务状态，任务状态请查询 /api/tasks）"""
    has_llm_config: bool
    graph_exists: bool
    node_count: int
    edge_count: int
    is_building: bool = False
    building_progress: int = 0
    is_full_build: bool = False
    last_built_at: Optional[str] = None


class BuildGraphRequest(BaseModel):
    """构建图谱请求"""
    note_ids: Optional[List[int]] = None
    rebuild: bool = False  # 是否重新构建（清空现有图谱）


class BuildGraphProgress(BaseModel):
    """构建图谱进度"""
    status: str
    progress: int
    message: str
    stats: Optional[Dict[str, Any]] = None
